execution/executor/src/components/apply_chunk_output.rs

```diff
@@ -191,106 +177,80 @@ impl ApplyChunkOutput {
-                transactions.drain(pos..).collect(),
-                transaction_outputs.drain(pos..).collect(),
-            )
-        } else if let Some(pos) = block_gas_limit_marker {
-            TransactionsWithParsedOutput::new(
-                transactions.drain(pos..).collect(),
-                transaction_outputs.drain(pos..).collect(),
-            )
-        } else {
-            TransactionsWithParsedOutput::new(vec![], vec![])
-        };
-
-        let state_checkpoint_to_add =
-            new_epoch_marker.map_or_else(|| append_state_checkpoint_to_block, |_| None);
-
-        let keeps_and_discards = transaction_outputs.iter().map(|t| t.status()).cloned();
-        let retries = repeat(TransactionStatus::Retry).take(to_retry.len());
 
-        let status = if state_checkpoint_to_add.is_some() {
-            keeps_and_discards
-                .chain(once(TransactionStatus::Keep(ExecutionStatus::Success)))
-                .chain(retries)
-                .collect()
-        } else {
-            keeps_and_discards.chain(retries).collect()
-        };
-
-        // Separate transactions with the Keep status out.
-        let (mut to_keep, to_discard) = itertools::zip_eq(transactions, transaction_outputs)
-            .partition::<Vec<(Transaction, ParsedTransactionOutput)>, _>(|(_, o)| {
-                matches!(o.status(), TransactionStatus::Keep(_))
-            });
-
-        // Append the StateCheckpoint transaction to the end of to_keep
-        if let Some(block_id) = state_checkpoint_to_add {
-            let state_checkpoint_txn = Transaction::StateCheckpoint(block_id);
-            let state_checkpoint_txn_output: ParsedTransactionOutput =
-                Into::into(TransactionOutput::new(
-                    WriteSet::default(),
-                    Vec::new(),
-                    0,
-                    TransactionStatus::Keep(ExecutionStatus::Success),
-                ));
-            to_keep.push((state_checkpoint_txn, state_checkpoint_txn_output));
+        let mut to_keep = vec![];
+        let mut to_discard = vec![];
+        let mut to_retry = vec![];
+
+        let mut found_epoch_ended_marker = false;
+        let mut found_state_checkpoint = false;
+        let mut found_retry = false;
+
+        for (transaction, transaction_output) in
+            itertools::zip_eq(transactions, transaction_outputs)
+        {
+            match transaction_output.status() {
+                TransactionStatus::Keep(_) => {
+                    if found_epoch_ended_marker {
+                        error!("Found Keep transaction after epoch ending marker, discarding.");
+                        to_discard.push((transaction, transaction_output));
+                        continue;
+                    }
+                    if found_state_checkpoint {
+                        error!("Found Keep transaction after state checkpoint, discarding.");
+                        to_retry.push((transaction, transaction_output));
```

Comment:
Why `to_retry` not `to_discard`? But anyway these cases shouldn't happen.

Reply:
that was the logic before. we find the position of the first Retry, and put into to_retry .

I can add a comment that this is from logic before. though since error doesn't show up in prod, maybe I can just put into to_discard. 

any opinions?

Reply:
Oh now I think they need to be `to_retry`. Any txn after reconfig or state checkpoint shouldn't be included in the blockchain state, so they should be `to_retry`.

Reply:
to_discard is also not included in the blockchain?

only to_keep goes on-chain

Reply:
@grao1991 to comfirm